File: civitasRUBY/lib/jugador.rb
#encoding: utf-8

require_relative 'jugador_especulador'

module Civitas
  class Jugador
    attr_reader :nombre, :numCasillaActual, :propiedades, :puedeComprar, :saldo
    attr_reader :salvoconducto, :encarcelado

    @@CasasMax        = 4
    @@CasasPorHotel   = 4
    @@HotelesMax      = 4
    @@PasoPorSalida   = 1000
    @@PrecioLibertad  = 200
    @@SaldoInicial    = 7500
    
    def initialize(nombre, encarcelado, numCasillaActual, puedeComprar, saldo, salvoconducto, propiedades)
      @nombre           = nombre
      @encarcelado      = encarcelado
      @numCasillaActual = numCasillaActual
      @puedeComprar     = puedeComprar
      @saldo            = saldo
      @salvoconducto    = salvoconducto
      @propiedades      = propiedades
    end
    
    #Constructor 1
    def self.new_nombre(nombre)
      new(nombre, false, 0, false, @@SaldoInicial, nil, Array.new)
    end
    
    #Constructor 2, de copia
    def self.new_otro(otro)
      new(otro.nombre, otro.encarcelado, otro.numCasillaActual, otro.puedeComprar, otro.saldo, otro.salvoconducto, otro.propiedades)
    end
    
    #Devuelve si el jugador debe ser encarcelado
    def debeSerEncarcelado
      
      if !@encarcelado
      
        if tieneSalvoconducto
          perderSalvoconducto
          Diario.instance.ocurreEvento("Jugador " + @nombre + " se libra de carcel por salvoconducto")
          
        else
          return true
        end
      end
      
      return false
    end

    public
    #toString de Jugador
    def toString(palabra="")
      cadena=  "{"+palabra+"\n\tNombre: #{@nombre}" +
        "\n\tCasilla: #{@numCasillaActual}"         +
        "\n\tSaldo: #{@saldo}"                      +
        "\n\tEstá encarcelado: "     + (@encarcelado ? "Sí" : "No")        +
        "\n\tTiene salvoconducto: "  + (@salvoconducto==nil ? "No" : "Sí") +
        "\n\tPropiedades:"
      
      for propiedad in @propiedades
        cadena+=propiedad.toString
      end
      cadena+="\n}"
      
      return cadena
    end
      
    #Devuelve si jugador está encarcelado
    def isEncarcelado
      return @encarcelado
    end

    #Realiza los pasos para encarcelar a jugador de ser posible. Devuelve si ha sido encarcelado
    def encarcelar(numCasillaCarcel)
      if debeSerEncarcelado
        moverACasilla(numCasillaCarcel)
        @encarcelado=true
        Diario.instance.ocurreEvento("Jugador " + @nombre + " ha sido enviado a la carcel")
      end

      return @encarcelado
    end

    #Otorga salvoconducto a jugador si no está encarcelado. Devuelve si se le otorga
    def obtenerSalvoconducto(sorpresa)
      if @encarcelado
        return false
      else

        @salvoconducto=sorpresa
        return true
      end
    end

    #Quita el salvoconducto a jugador
    def perderSalvoconducto
      @salvoconducto.usada
      @salvoconducto=nil      
    end

    #Devuelve si el jugador posee salvoconducto
    def tieneSalvoconducto
      if @salvoconducto != nil
        return true
      end
      
      return false
    end

    #Devuelve si el jugador puede comprar casillas
    def puedeComprarCasilla
      @puedeComprar= !@encarcelado
      return @puedeComprar
    end

    #Realiza el proceso de pago
    def paga(cantidad)
      return modificarSaldo(cantidad*-1)
    end

    #Realiza el proceso de pago de impuestos de ser posible. Devuelve si es posible
    def pagaImpuesto(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end
    end

    #Realiza el proceso de pagar el alquiler de ser posible. Devuelve si es posible
    def pagaAlquiler(cantidad)
      if @encarcelado
        return false
      else
        return paga(cantidad)
      end
    end

    #Realiza el proceso de recibir dinero de ser posible. Devuelve si es posible
    def recibe(cantidad)
      if @encarcelado
        return false
      else
        return modificarSaldo(cantidad)
      end
    end

    #Incrementa el saldo
    def modificarSaldo(cantidad)
      @saldo+=cantidad
     
      Diario.instance.ocurreEvento("El jugador "+@nombre+(cantidad>0 ? " añade":" retira")+cantidad.to_s+" a su saldo")
      
      return true
    end

    #Mueve al jugador a la casilla indicada
    def moverACasilla(numCasilla)
      if !@encarcelado
      
        @numCasillaActual=numCasilla
        
        puedeComprar=false
       
        Diario.instance.ocurreEvento("El jugador "+@nombre+" ha sido trasladado a la casilla "+numCasilla.to_s)
        return true
      end
      return false
    end
    
    #Devuelve si el jugador puede gastar la cantidad indicada
    def puedoGastar(precio)
      if !@encarcelado && @saldo>precio
        return true
      end

      return false
    end

    #Realiza todo el procedimiento de venta de una propiedad ip, de ser posible. Devuelve si es posible
    def vender(ip)
      if !@encarcelado && existeLaPropiedad(ip)
        if @propiedades.at(ip).vender(self)
          Diario.instance.ocurreEvento("Jugador "+@nombre+" ha vendido propiedad "+@propiedades.at(ip).nombre)
          @propiedades.delete_at(ip)
          
          return true
        end
      end
      
      return false
    end

    #Devuelve si el jugador tiene propiedades
    def tieneAlgoQueGestionar
      tiene = false
      if @propiedades.size > 0
        tiene = true
      end
      return tiene
    end
    
    #Devuelve si tiene dinero para salir de la cárcel
    def puedeSalirCarcelPagando
      return @saldo>@@PrecioLibertad
    end

    #Saca de la carcel a jugador de ser posible. Devuelve si es posible
    def salirCarcelPagando
      if puedeSalirCarcelPagando
        paga(@@PrecioLibertad)
        @encarcelado=false
        Diario.instance.ocurreEvento("El jugador "+@nombre+" ha salido de la carcel pagando la fianza")

        return true
      end

      return false
    end

    #Saca de la carcela al jugador si la tirada es favorable. Devuelve si la tirada fue favorable
    def salirCarcelTirando
      if Dado.instance.tirar>=5 #reglas del juego
        @encarcelado=false
        Diario.instance.ocurreEvento("El jugador "+@nombre+" ha salido de la carcel mediante tirada de dado")
        return true
      end

      return false
    end

    #Realiza el proceso por el que se le ingresa dinero al jugador por pasar por salida
    def pasaPorSalida
      modificarSaldo(@@PasoPorSalida)
      Diario.instance.ocurreEvento("El jugador "+@nombre+" pasa por salida y se lleva "+@@PasoPorSalida.to_s)

      return true
    end

    #compareTo de Jugador
    def <=> (otro)
      @saldo <=> otro.saldo
    end

    #Devuelve la cantidad de casas y hoteles de jugador
    def cantidadCasasHoteles
      total=0
        
      for prop in @propiedades
        total += prop.numCasas
        total += prop.numHoteles
      end
        
      return total
    end

    #Devuelve si el jugador está en bancarrota
    def enBancarrota        
      if @saldo < 0
        return true
      else
        return false
      end
    end

    #Devuelve si existe la propiedad con índice ip
    def existeLaPropiedad(ip)
      existe = false
        
      if @propiedades.size>0  &&  propiedades.at(ip)!=nil
        if @propiedades.at(ip).propietario==self
          existe=true
        end
      end
        
      return existe
    end

    #Devuelve si el jugador puede edificar una casa en la Propiedad propiedad
    def puedoEdificarCasa(propiedad)
      if propiedad.numCasas<@@CasasMax
        return true
      else
        return false
      end
      
    end

    #Devuelve si el jugador puede edificar un hotel en la Propiedad propiedad
    def puedoEdificarHotel(propiedad)

      if propiedad.numHoteles<@@HotelesMax
        return true
      else
        return false
      end
      
    end
    
    #Devuelve si el jugador está encarcelado
    def isEncarcelado
      @encarcelado
    end
    
    #Cancela la hipoteca con índice ip
    def cancelarHipoteca(ip)
      if @encarcelado
        return false
      else
        if existeLaPropiedad(ip)
          propiedad   = @propiedades.at(ip)
          cantidad    = propiedad.importeCancelarHipoteca

          if puedoGastar(cantidad)
            propiedad.cancelarHipoteca(self)
            Diario.instance.ocurreEvento("El jugador "+@nombre+" cancela la hipoteca de la propiedad "+ip.to_s)

            return true
          end
        end
      end

      return false
    end
    
    #Comprar una propiedad de ser posible. Devuelve si ha sido posible
    def comprar(titulo)
      if (@encarcelado)
        return false
      end
      
      if (@puedeComprar)
        if (puedoGastar(titulo.precioCompra))
           
          if (titulo.comprar(self))
            @propiedades.push(titulo)
            Diario.instance.ocurreEvento("El jugador "+@nombre+" compra la propiedad "+titulo.nombre)
          
            @puedeComprar = false;
          end
        
          return true
        end
      end
      
      return false
    end
    
    #Construye un hotel en la propiedad ip
    def construirHotel (ip)
      seConstruye = false
      if (@encarcelado)
        return seConstruye
      end
      
      if (existeLaPropiedad(ip))
        propiedad = @propiedades[ip]
        
        if (puedoEdificarHotel(propiedad))
          seConstruye = propiedad.construirHotel(self)
          propiedad.derruirCasas(@@CasasPorHotel, self)
        end
        
        Diario.instance.ocurreEvento("El jugador "+@nombre+" construye un hotel en la propiedad "+propiedad.nombre)
      end
      
      return seConstruye
    end
    
    def construirCasa (ip)
      seConstruye = false
      
      if (@encarcelado)
        return seConstruye
      end
      
      if (existeLaPropiedad(ip))
        propiedad = @propiedades[ip]
        if (puedoEdificarCasa(propiedad))
          seConstruye = propiedad.construirCasa(self)
          if (seConstruye)
            Diario.instance.ocurreEvento("El jugador " + @nombre + " construye una casa en la propiedad " + propiedad.nombre)
          end
        end
      end
      
      return seConstruye
    end
    
    #Hipoteca la propiedad de índice ip de ser posible. Devuelve si fue posible
    def hipotecar(ip)
      if !@encarcelado  &&  existeLaPropiedad(ip)
        if propiedades.at(ip).hipotecar(self)
          Diario.instance.ocurreEvento("El jugador "+@nombre+ " hipoteca la propiedad "+ip.to_s)
          
          return true
        end
      end
      
      return false
    end
        
    #ESTE MÉTODO NO ES PARTE DEL GUIÓN. IMPLEMENTACIÓN PARA SOLUCIONAR VistasTextual::gestionar()
    #PARA EVITAR TENER QUE DEVOLVER TODO 'propiedades', devolvemos solo lo imprescindible
    #Devuelve el nombre de las propiedades
    def getNombrePropiedades
      nombres = Array.new;
        
      for propiedad in propiedades
        nombres<<propiedad.nombre;
      end
            
      return nombres;
    end
    
    #Devuelve jugador convertido a especulador
    def aEspeculador
      if !esEspeculador
        return JugadorEspeculador.new_otro(self, 2000)
      end
    end
    
    #ESTE MÉTODO NO ES PARTE DEL GUIÓN.
    # - Este método es para evitar tener que hacer públicas las propiedades
    #Jugador cede propiedades a otroJugador
    def cedePropiedadesA(otroJugador)
      for propiedad in @propiedades
        propiedad.actualizaPropietarioPorConversion(otroJugador);
      end
    end
    
    def self.casasMax
      return @@CasasMax
    end
    
    def self.casasPorHotel
      return @@CasasPorHotel
    end
    
    def self.hotelesMax
      return @@HotelesMax
    end
    
    def self.pasoPorSalida
      return @@PasoPorSalida
    end
    
    def self.precioLibertad
      return @@PrecioLibertad
    end
    
    def self.saldoInicial
      return @@SaldoInicial
    end
    
    #Devuelve si el jugador es especulador
    def esEspeculador
      return false
    end
  end
end

File: civitasJAVA/src/civitas/Jugador.java
package civitas;

import java.util.ArrayList;


public class Jugador implements Comparable<Jugador> {

    protected static int CasasMax         = 4;
    protected static int CasasPorHotel    = 4;
    protected static int HotelesMax       = 4;
    private int numCasillaActual          = -1;
    protected static float PasoPorSalida  = 1000;
    protected static float PrecioLibertad = 200;
    private static float SaldoInicial     = 7500;
    private boolean puedeComprar          = true;
    protected boolean encarcelado         = false;
    private String nombre;
    private float saldo;
    private ArrayList<TituloPropiedad> propiedades;
    private SorpresaSalvoconducto salvoconducto = null;

    //Constructor 1
    Jugador(String _nombre) {
        nombre = _nombre;
        numCasillaActual = 0;
        puedeComprar = false;
        encarcelado = false;
        saldo = SaldoInicial;
        propiedades = new ArrayList<TituloPropiedad>();
        salvoconducto = null;

    }

    //Constructor 2, de copia
    protected Jugador(Jugador otro) {
        this.nombre = otro.nombre;
        this.encarcelado = otro.encarcelado;
        this.numCasillaActual = otro.numCasillaActual;
        this.puedeComprar = otro.puedeComprar;
        this.saldo = otro.saldo;
        this.propiedades = otro.propiedades;
        this.salvoconducto = otro.salvoconducto;
    }

    //Devuelve si el jugador debe ser encarcelado
    protected boolean debeSerEncarcelado() {
        if (!encarcelado) {
            if (tieneSalvoconducto()) {
                perderSalvoconducto();
                Diario.getInstance().ocurreEvento("Jugador " + nombre + " se libra de carcel por salvoconducto");

                return false;
            }
        }

        return true;
    }

    //Realiza los pasos para encarcelar a jugador de ser posible. Devuelve si ha sido encarcelado
    boolean encarcelar(int numCasillaCarcel) {

        if (debeSerEncarcelado()) {
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            Diario.getInstance().ocurreEvento("Jugador " + nombre + " ha sido enviado a la carcel");
        }
        return encarcelado;
    }

    //Otorga salvoconducto a jugador si no está encarcelado. Devuelve si se le otorga
    boolean obtenerSalvoconducto(SorpresaSalvoconducto sorpresa) {
        if (encarcelado) {
            return false;
        } else {
            salvoconducto = sorpresa;
            return true;
        }

    }

    //CAMBIO RESPECTO A GUIÓN
    // -Eliminar visibilidad private para poder llamar a este método desde JugadorEspeculador
    //Quita el salvoconducto a jugador
    void perderSalvoconducto() {
        salvoconducto.usada();
        salvoconducto = null;
    }

    //Devuelve si el jugador posee salvoconducto
    boolean tieneSalvoconducto() {

        if (salvoconducto == null) {
            return false;
        } else {
            return true;
        }
    }

    //Devuelve si el jugador puede comprar casillas
    boolean puedeComprarCasilla() {
        puedeComprar = !encarcelado;

        return puedeComprar;
    }

    //Realiza el proceso de pago
    boolean paga(float cantidad) {
        return modificarSaldo(cantidad * - 1);
    }

    //Realiza el proceso de pago de impuestos de ser posible. Devuelve si es posible
    boolean pagaImpuesto(float cantidad) {

        if (encarcelado) 
            return false;
        else
            return paga(cantidad);
    }

    //Realiza el proceso de pagar el alquiler de ser posible. Devuelve si es posible
    boolean pagaAlquiler(float cantidad) {
        boolean resultado = false;

        if (!encarcelado)
            resultado = paga(cantidad);

        return resultado;
    }

    //Realiza el proceso de recibir dinero de ser posible. Devuelve si es posible
    boolean recibe(float cantidad) {
        boolean resultado = false;
        if (!encarcelado) {
            resultado = modificarSaldo(cantidad);

        }
        return resultado;
    }

    //Incrementa el saldo
    boolean modificarSaldo(float cantidad) {
        boolean verdadero = true;

        saldo = saldo + cantidad;
        Diario.getInstance().ocurreEvento("El jugador " + nombre + " tiene de saldo " + saldo);

        return verdadero;
    }

    //Mueve al jugador a la casilla indicada
    boolean moverACasilla(int numCasilla) {
        if (!encarcelado) {
            if (numCasilla < 0) {
                System.out.print("ADJNSKJNS");
            }
            numCasillaActual = numCasilla;
            
            puedeComprar = false;
            
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha sido trasladado a la casilla " + numCasilla);

            return true;
        }

        return false;
    }

    //MODIFICACIÓN DE GUIÓN
    //  -Eliminar visibilidad private para poder llamar al método desde JugadorEspeculador
    //Devuelve si el jugador puede gastar la cantidad indicada
    boolean puedoGastar(float precio) {

        if (!encarcelado && saldo > precio) {
            return true;
        } else {
            return false;
        }
    }

    //Realiza todo el procedimiento de venta de una propiedad ip, de ser posible. Devuelve si es posible
    boolean vender(int ip) {

        if (!encarcelado && existeLaPropiedad(ip)) {
            if (propiedades.get(ip).vender(this)) {
                Diario.getInstance().ocurreEvento("Jugador " + nombre + " ha vendido propiedad " + propiedades.get(ip).getNombre());
                propiedades.remove(ip);

                return true;
            }
        }

        return false;
    }

    //Devuelve la cantidad de casas y hoteles de jugador
    int cantidadCasasHoteles() {
        int total = 0;

        //llamamos a la función cantidadCasasHoteles de TituloPropiedad
        for (int i = 0; i < propiedades.size(); i++) {
            total = propiedades.get(i).cantidadCasasHoteles();
        }

        return total;
    }

    //compareTo de Jugador
    public int compareTo(Jugador otro) {

        //se encarga de comparar el resultado de dos jugadores
        int resultado = Float.compare(saldo, otro.saldo);

        return resultado;
    }

    //Devuelve si el jugador está en bancarrota
    boolean enBancarrota() {
        boolean enBancarrota = false;

        if (saldo < 0) {
            enBancarrota = true;
        }

        return enBancarrota;
    }

    //Devuelve si existe la propiedad con índice ip
    private boolean existeLaPropiedad(int ip) {
        boolean existe = false;

        if (!propiedades.isEmpty()) {
            if (propiedades.get(ip).getPropietario() == this) {
                existe = true;
            }
        }

        return existe;
    }

    //CAMBIO RESPECTO A GUIÓN
    // - Elimino visibilidad private para poder acceder desde JugadorEspeculador
    //Get de casasMax
    int getCasasMax() {
        return CasasMax;
    }

    //CAMBIO RESPECTO A GUIÓN
    // - Elimino visibilidad private para poder acceder desde JugadorEspeculador
    //Get de HotelesMax
    int getHotelesMax() {
        return HotelesMax;
    }

    //Get de CasasPorHotel
    int getCasasPorHotel() {
        return CasasPorHotel;
    }

    //Get de nombre
    protected String getNombre() {
        return nombre;
    }

    //Get de numCasillaActual
    int getNumCasillaActual() {
        return numCasillaActual;
    }

    //Get de PrecioLibertad
    float getPrecioLibertad() {
        return PrecioLibertad;
    }

    //Get de PasoPorSalida
    float PasoPorSalida() {
        return PasoPorSalida;
    }

    //Get de propiedades
    protected ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    //Get de puedeComprar
    boolean getPuedeComprar() {
        return puedeComprar;
    }

    //Get de saldo
    protected float getSaldo() {
        return saldo;
    }

    //Devuelve si jugador está encarcelado
    public boolean isEncarcelado() {
        return encarcelado;
    }

    //Realiza el proceso por el que se le ingresa dinero al jugador por pasar por salida
    boolean pasaPorSalida() {
        modificarSaldo(PasoPorSalida);

        Diario.getInstance().ocurreEvento("El jugador " + nombre + " pasa por salida y se lleva " + PasoPorSalida + "€");

        return true;
    }

    //Devuelve si tiene dinero para salir de la cárcel
    private boolean puedeSalirCarcelPagando() {
        return saldo > PrecioLibertad;
    }

    //Devuelve si el jugador puede edificar una casa en la Propiedad propiedad
    boolean puedoEdificarCasa(TituloPropiedad propiedad) {
        return (propiedad.getNumCasas() < CasasMax ? true : false);
    }

    //Devuelve si el jugador puede edificar un hotel en la Propiedad propiedad
    boolean puedoEdificarHotel(TituloPropiedad propiedad) {
        return (propiedad.getNumHoteles() < HotelesMax ? true : false);
    }

    //Saca de la carcel a jugador de ser posible. Devuelve si es posible
    boolean salirCarcelPagando() {

        if (puedeSalirCarcelPagando()) {
            encarcelado = false;
            paga(PrecioLibertad);
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha salido de la carcel pagando la fianza");
            return true;
        }
        return false;
    }

    //Saca de la carcela al jugador si la tirada es favorable. Devuelve si la tirada fue favorable
    boolean salirCarcelTirando() {
        boolean salgo = Dado.getInstance().tirar() >= 5;
        if (salgo) {
            encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " ha salido de la carcel mediante tirada de dado");
            salgo = true;
        }
        return salgo;
    }

    //Devuelve si el jugador tiene propiedades
    boolean tieneAlgoQueGestionar() {
        boolean tiene = false;

        if (!propiedades.isEmpty()) {
            tiene = true;
        }

        return tiene;
    }

    //toString de Jugador
    @Override
    public String toString() {
        return "Jugador :"                   + nombre
                + "\n\tCasilla = "           + numCasillaActual
                + "\n\tPuede comprar =  "    + puedeComprar
                + "\n\tEstá encarcelado =  " + encarcelado
                + "\n\tSaldo = "             + saldo
                + "\n\tTiene salvoconducto " + salvoconducto
                + "\n\tPropiedades = "       + getPropiedades().toString();

    }

    //Cancela la hipoteca con índice ip
    boolean cancelarHipoteca(int ip) {
        if (encarcelado) {
            return false;
        }

        if (existeLaPropiedad(ip)) {
            TituloPropiedad propiedad = propiedades.get(ip);
            float cantidad = propiedad.getImporteCancelarHipoteca();

            if (puedoGastar(cantidad)) {
                propiedad.cancelarHipoteca(this);
                Diario.getInstance().ocurreEvento("El jugador " + nombre + " cancela la hipoteca de la propiedad " + ip);
                return true;
            }
        }

        return false;
    }

    //Hipoteca la propiedad de índice ip de ser posible. Devuelve si fue posible
    boolean hipotecar(int ip) {
        boolean result = false;

        if (encarcelado) {
            return result;
        }

        if (existeLaPropiedad(ip)) {

            TituloPropiedad propiedad = propiedades.get(ip);
            result = propiedad.hipotecar(this);
        }

        if (result) {
            Diario.getInstance().ocurreEvento("El jugador " + nombre + " hipoteca la propiedad " + ip);
        }

        return result;
    }

    //Comprar una propiedad de ser posible. Devuelve si ha sido posible
    boolean comprar(TituloPropiedad titulo) {
        if (encarcelado) {
            return false;
        }

        if (puedeComprar) {
            if (puedoGastar(titulo.getPrecioCompra())) {
                if (titulo.comprar(this)) {
                    propiedades.add(titulo);
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + " compra la propiedad " + titulo.toString());
                    puedeComprar = false;
                }

                return true;
            }
        }

        return false;
    }

    //Construye una casa
    boolean construirCasa(int ip) {
        if (encarcelado)
            return false;

        if (existeLaPropiedad(ip)) {
            TituloPropiedad propiedad = propiedades.get(ip);
           
            if (puedoEdificarCasa(propiedad)) {
                if (propiedad.construirCasa(this)) {
                    Diario.getInstance().ocurreEvento("El jugador " + nombre + " construye casa en la propiedad " + ip);
                    
                    return true;
                }
            }
        }

        return false;
    }

    //Construir un hotel
    boolean construirHotel(int ip) {
        if (this.isEncarcelado()) 
            return false;

        if (existeLaPropiedad(ip)) {
            TituloPropiedad propiedad = propiedades.get(ip);
            
            if (puedoEdificarHotel(propiedad)) {
                propiedad.construirHotel(this);
                propiedad.derruirCasas(CasasPorHotel, this);
                Diario.getInstance().ocurreEvento("El jugador " + getNombre() + " construye hotel en la propiedad " + ip);
            
                return true;
            }
        }

        return false;
    }

    //ESTE MÉTODO NO ES PARTE DEL GUIÓN. IMPLEMENTACIÓN PARA SOLUCIONAR VistasTextual::gestionar()
    //PARA EVITAR TENER QUE DEVOLVER TODO 'propiedades', devolvemos solo lo imprescindible
    //Devuelve el número de propiedades
    public ArrayList<String> getNombrePropiedades() {
        ArrayList<String> nombres = new ArrayList<String>();

        for (TituloPropiedad propiedad : propiedades) {
            nombres.add(propiedad.getNombre());
        }

        return nombres;
    }

    //ESTE MÉTODO NO ES PARTE DEL GUIÓN.
    // - Este método es para evitar tener que hacer públicas las propiedades
    //Jugador cede propiedades a otroJugador
    public void cedePropiedadesA(Jugador otroJugador) {
        for (TituloPropiedad propiedad : propiedades) {
            propiedad.actualizaPropietarioPorConversion(otroJugador);
        }
    }

    //Devuelve jugador convertido a especulador
    public JugadorEspeculador aEspeculador() {
        return new JugadorEspeculador(this, 2000);
    }
    
    //Devuelve si el jugador es especulador
    public boolean esEspeculador()
    {
        return false;
    }
}
